migrate humble to jazzy

WHongLin opened the pull request:
Hi, I have migrated franka_ros2 from humble to jazz. 

franka_gazebo is available now~

WHongLin committed:
refactor: migrate to jazzy

example controllers changes in implicit type conversion: set_value()

BREAKING CHANGE:
  - for jazzy function interfaces changes: ControllerInterfaceBase::init()
  - for jazzy function interfaces changes: ControllerInterfaceBase::init()
  - migrate ignition gazebo to gz-sim-vendor,remove franka_ign_ros2_control

WHongLin committed:
refactor: adapt to CommandInterface are change to set_value() get_optional()

set_value() returns a bool instead of void, and is declared as [[nodiscard]]
get_value() is declared as [[deprecated]], use get_optional() instead

WHongLin committed:
fix: remove sapce in franka_gz_ros2_control folder name and format file

Commit:
chore: franka_bringup python file flake8 check

Commit:
fix: package.xml depend franka_gz_ros2_control

Commit:
chore: disable some new checks from clang-tidy

Commit:
fix: controllers configure failed

franka_robot_state_broacaset use new api to get urdf
model_example_controller auto declare parameters before use
joint_impedance_with_ik_example_controller declare parameters berfore use

Commit:
chore: code format check

Commit:
chore: use --controller-ros-args to remap topics

Commit:
refactor: modify franka_fr3_moveit_config launch configuration

add joint limits for response adapter
change lma kinematic solver to kdl
modify config format for move_group

Commit:
docker: change the dependency to ros-jazzy

Commit:
refactor: (unavailable)plugin loading issue in franka_gz_ros2_control

Commit:
fix: plugin loading issue in franka_gazebo

inherits a new ResourceManager to control the init process
add ros_gz_bridge in launch file

AndreasKuhner commented on franka_bringup/utils/launch_utils.py:
```diff
@@ -13,11 +13,12 @@
 
 def load_yaml(file_path):
     if not os.path.exists(file_path):
-        raise FileNotFoundError(f"File not found: {file_path}")
```
Are all the changes from double quotes to single quotes necessary?

WHongLin replied:
Hi, double quotes cause a failure on my side when I run `colcon test`:
- flake8
  <<< failure message
    -- run_test.py: invoking following command in '/home/franka/work/franka_ros2_ws/src/franka_bringup':
     - /opt/ros/jazzy/bin/ament_flake8 --xunit-file /home/franka/work/franka_ros2_ws/build/franka_bringup/test_results/franka_bringup/flake8.xunit.xml
                        FindPackageShare('franka_bringup'), 'config', "controllers.yaml",
                                                                      ^
    1     Q000 Double quotes found but single quotes preferred
    
    ./launch/example.launch.py:108:67: Q000 Double quotes found but single quotes preferred
    
    
    5 files checked
    1 errors
    
    'Q'-type errors: 1

AndreasKuhner commented on franka_example_controllers/src/cartesian_elbow_example_controller.cpp:
```diff
@@ -50,12 +51,22 @@ controller_interface::return_type CartesianElbowExampleController::update(
     initial_pose_configuration_ = franka_cartesian_pose_->getCurrentPoseMatrix();
 
-    initial_robot_time_ = state_interfaces_.back().get_value();
+    if (auto time_op = state_interfaces_.back().get_optional<double>()) {
```
Did `state_interfaces` change s.t. they always return now optional values..?

WHongLin replied:
The `double get_value()` function in state/command interface has been marked as deprecated and will be removed in next ROS release. However, compilation tools currently issue warnings about this, so I replace it with `get_optional()`.
get_optional() first checks if the data is of double type, and then performs the same operations for double types as get_value().
This has also led to some of the "discontinuous instructions" issues you mentioned, and I will further consider this problem.

AndreasKuhner commented on franka_example_controllers/src/joint_impedance_with_ik_example_controller.cpp:
```diff
@@ -66,9 +67,15 @@ void JointImpedanceWithIKExampleController::update_joint_states() {
-    joint_positions_current_[i] = position_interface.get_value();
-    joint_velocities_current_[i] = velocity_interface.get_value();
-    joint_efforts_current_[i] = effort_interface.get_value();
+    // joint_positions_current_[i] = position_interface.get_value();
```
I guess outcommented code can be removed?

WHongLin replied:
Thanks~ I will check and delete these commented code.

WHongLin replied:
Thanks~ I will check and delete these commented code.

AndreasKuhner commented on franka_example_controllers/src/joint_impedance_with_ik_example_controller.cpp:
```diff
@@ -66,9 +67,15 @@ void JointImpedanceWithIKExampleController::update_joint_states() {
+    // joint_efforts_current_[i] = effort_interface.get_value();
+
+    joint_positions_current_[i] =
+        position_interface.get_optional().value_or(joint_positions_current_[i]);
```
Wouldn't this alter the behavior/causes errors in case of no available error?

If you use the same position twice, it sounds like a very discontinuous signal.

AndreasKuhner commented on franka_example_controllers/src/move_to_start_example_controller.cpp:
```diff
@@ -48,6 +48,13 @@ MoveToStartExampleController::state_interface_configuration() const {
+    for (auto& command_interface : command_interfaces_) {
+      ret = command_interface.set_value(0.0);
+    }
+    return ret;
```
Wouldn't you need to do a `ret = ret && command_interface.set_value(0.0);`? Otherwise, you would only return the last 'success'

WHongLin replied:
Yes, it should be like this.

AndreasKuhner commented on franka_example_controllers/src/move_to_start_example_controller.cpp:
```diff
@@ -59,12 +66,13 @@ controller_interface::return_type MoveToStartExampleController::update(
     for (int i = 0; i < 7; ++i) {
-      command_interfaces_[i].set_value(tau_d_calculated(i));
+      if (!command_interfaces_[i].set_value(tau_d_calculated(i))) {
+        clear_tau();
```
In case of 'setting was not successful' -> If you set the torque to 0, wouldn't that cause a big jump?

AndreasKuhner commented on franka_example_controllers/src/move_to_start_example_controller.cpp:
```diff
@@ -131,8 +139,11 @@ void MoveToStartExampleController::updateJointStates() {
-    q_(i) = position_interface.get_value();
-    dq_(i) = velocity_interface.get_value();
+    // q_(i) = position_interface.get_value();
+    // dq_(i) = velocity_interface.get_value();
```
outcommented code

AndreasKuhner commented on franka_fr3_moveit_config/config/fr3_joint_limits.yaml:
Is this now needed with jazzy? Or might this then something which can be added to our URDFs?

WHongLin replied:
yes~The OMPL planning pipeline response adapter AddTimeOptimalParameterization now requires acceleration parameters. These parameters can be transferred via URDF or YAML configuration.

logs:
[move_group-3] [INFO] [1754890049.290263178] [move_group]: Calling Planner 'OMPL'
[move_group-3] [INFO] [1754890049.306819322] [move_group]: Calling PlanningResponseAdapter 'AddTimeOptimalParameterization'
[move_group-3] [ERROR] [1754890049.306884577] [moveit.moveit.core.time_optimal_trajectory_generation]: No acceleration limit was defined for joint fr3_joint1! You have to define acceleration limits in the URDF or joint_limits.yaml
[move_group-3] [ERROR] [1754890049.306914429] [moveit.moveit.ros.add_time_optimal_parameterization]: Response adapter 'AddTimeOptimalParameterization' failed to generate a trajectory.
[move_group-3] [ERROR] [1754890049.306935850] [move_group]: PlanningResponseAdapter 'AddTimeOptimalParameterization' failed with error code FAILURE
[move_group-3] [INFO] [1754890049.306975749] [moveit.moveit.ros.move_group.move_action]: FAILURE
[rviz2-1] [INFO] [1754890049.307297852] [rviz2.moveit.ros.move_group_interface]: Plan and Execute request aborted
[rviz2-1] [ERROR] [1754890049.308094144] [rviz2.moveit.ros.move_group_interface]: MoveGroupInterface::move() failed or timeout reached

AndreasKuhner commented on franka_fr3_moveit_config/config/kinematics.yaml:
How would change this the overall behavior?

WHongLin replied:
Use different IK solvers: kdl_kinematics_plugin uses KDL::ChainIkSolverVelMimicSVD, and lma_kinematics_plugin uses KDL::ChainIkSolverPos_LMA. Additionally, the Jazzy version of MoveIt does not provide an LMA one.

AndreasKuhner commented on franka_gazebo/franka_gz_ros2_control/src/gz_ros2_control_plugin.cpp:
```diff
@@ -40,18 +41,89 @@
+
+private:
+  std::shared_ptr<rclcpp::Node> node_;
+  gz::sim::EntityComponentManager & ecm_;
```
Rename please: `ecm_` -> `entity_component_manager_`

Can we get this value only as ref? or could we get it as `shared_ptr`? Storing a ref is always very risky...

WHongLin replied:
The ecm_ is created by GZROS2ControlPlugin and passed to GZResourceManager during its lifecycle. GZResourceManager only stores it temporarily and passes it to GZSystem during initialization; ultimately, it is stored as a pointer in GZSystem. Both the lifecycles of GZSystem and GZResourceManager are shorter than that of GZROS2ControlPlugin.
Throughout the entire parameter passing process, references of ecm_ are used, and ultimately, it is stored as a pointer in GZSystem. so I use a ref too~

AndreasKuhner commented on franka_gazebo/franka_gz_ros2_control/src/gz_ros2_control_plugin.cpp:
```diff
@@ -40,18 +41,89 @@
 {
-//////////////////////////////////////////////////
-class IgnitionROS2ControlPluginPrivate
+class GZResourceManager : public hardware_interface::ResourceManager
```
Is this class only living in the implementation (aka .cpp file)?

WHongLin replied:
Yes, this class is only used in this file. The GZResourceManager can only be created when we use the Gazebo simulation plugin.

AndreasKuhner commented on franka_gazebo/franka_gz_ros2_control/src/gz_ros2_control_plugin.cpp:
```diff
@@ -40,18 +41,89 @@
+
+  bool load_and_initialize_components(
+    const std::string & urdf, const unsigned int update_rate) override
+  {
```
I am a bit surprise by the formatting -> Usually the `{` is at the end of the last line. Do you use clang-format?

WHongLin replied:
I'm sorry. After checking my development environment, I found that a plugin has changed the formatting tool in VSCode from clangd to another one. I will check this issue again.

AndreasKuhner commented on franka_hardware/include/franka_hardware/franka_hardware_interface.hpp:
```diff
@@ -21,7 +21,7 @@
 #include <vector>
 
-#include <hardware_interface/visibility_control.h>
+// #include <hardware_interface/visibility_control.h>
```
commented line

AndreasKuhner commented on franka_hardware/test/franka_hardware_interface_test.cpp:
```diff
@@ -189,7 +189,8 @@ TEST_F(
       ASSERT_EQ(states[i].get_name(), arm_id + "_" + joint_name + "/" + k_effort_controller);
     }
-    ASSERT_EQ(states[i].get_value(), 0.0);
+    // ASSERT_EQ(states[i].get_value(), 0.0);
```
commented line

AndreasKuhner commented on franka_robot_state_broadcaster/src/franka_robot_state_broadcaster.cpp:
```diff
@@ -23,7 +23,7 @@
 #include <rclcpp/clock.hpp>
 #include <rclcpp/qos.hpp>
-#include <rclcpp/qos_event.hpp>
+// #include <rclcpp/qos_event.hpp>
```
commented line

AndreasKuhner commented on franka_semantic_components/src/franka_semantic_component_interface.cpp:
```diff
@@ -69,7 +69,8 @@ std::vector<double> FrankaSemanticComponentInterface::get_values_state_interface
   // insert all the state_interface_values
   for (const auto& state_interface : state_interfaces_) {
-    state_interface_values.emplace_back(state_interface.get().get_value());
+    // state_interface_values.emplace_back(state_interface.get().get_value());
```
commented line

AndreasKuhner commented on franka_semantic_components/src/franka_semantic_component_interface.cpp:
```diff
@@ -78,7 +79,8 @@ std::vector<double> FrankaSemanticComponentInterface::get_values_command_interfa
   // insert all the command_interface_values
   for (const auto& command_interface : command_interfaces_) {
-    command_interface_values.emplace_back(command_interface.get().get_value());
+    // command_interface_values.emplace_back(command_interface.get().get_value());
```
commented line

Jelmerdw commented:
@WHongLin Thanks a lot for your contribution!

We tried to use Jazzy before, but then we encountered the communication_constraints_violation error as mentioned here:

https://github.com/frankarobotics/franka_ros2/issues/108

We tried your PR and still faced the same problem (while using system image 5.8). We now tried your PR with system image 5.8.1, and this seems to work well. It looks like the problem now only happens when we try to move the arm too quickly after launching or when we launch the robot quickly again after a previous launch.

Commit:
chore: rm commented code and apply franka test options for franka_gazebo

Commit:
fix: franka_fr3_moveit_config launch file

fix topic remap in moveit.launch.py
modify kinematics config in move_group.launch.py

BarisYazici commented on franka_example_controllers/src/cartesian_orientation_example_controller.cpp:
```diff
@@ -47,12 +47,22 @@ controller_interface::return_type CartesianOrientationExampleController::update(
     std::tie(orientation_, position_) =
         franka_cartesian_pose_->getCurrentOrientationAndTranslation();
-    initial_robot_time_ = state_interfaces_.back().get_value();
+    if (auto time_op = state_interfaces_.back().get_optional<double>()) {
```
this seems like a duplicate of what we have in line [60](https://github.com/frankarobotics/franka_ros2/pull/170/files#diff-8a64d85addac766525d879ca83fe3663006d1907886f2667d050f4839c1f3dd0R60), could we somehow refactor this to a function and call that instead.

WHongLin replied:
Yeah, I will extract "auto time_op = state_interfaces_.back().get_optional<double>()" out of the if(initialization_flag_)...else section, then assign `time_op` to different time variables according to different branches.

BarisYazici commented on franka_example_controllers/src/joint_impedance_with_ik_example_controller.cpp:
```diff
@@ -179,15 +196,23 @@ controller_interface::return_type JointImpedanceWithIKExampleController::update(
 }
 
 CallbackReturn JointImpedanceWithIKExampleController::on_init() {
+  auto_declare("arm_id", "fr3");
```
Why are theses auto declares necessary? Did we have them before?

WHongLin replied:
Hi~ I added these auto declares for several controllers, because Nodes don't accept those undeclared parameters .Another method is to set NodeOptions::allow_undeclared_parameters(true).

Commit:
refactor: optimize get/set in example controller
  optimize duplicate code for getting robot time
  remove discontinous get/set commands

BarisYazici commented on franka_semantic_components/src/franka_robot_model.cpp:
```diff
@@ -60,8 +60,20 @@ void FrankaRobotModel::initialize() {
       franka_model_interface != state_interfaces_.end()) {
-    robot_model_ = bit_cast<franka_hardware::Model*>((*franka_model_interface).get().get_value());
-    robot_state_ = bit_cast<franka::RobotState*>((*franka_state_interface).get().get_value());
+    if (auto model_ptr = (*franka_model_interface).get().get_optional()) {
```
can you refactor the duplicate code?

BarisYazici commented:
can you also migrate the github ci workflow to jazzy branch?
https://github.com/frankarobotics/franka_ros2/blob/humble/.github/workflows/ci.yml

BarisYazici commented:
I noticed that in the Jazzy when a controller returns Error in any state, it's immediately shutdown(maybe it was the default case also for the humble not sure). But can we update this [configuration](https://github.com/frankarobotics/franka_ros2/blob/a739caa35161035f3a4505fb077068e9adf51172/franka_bringup/config/controllers.yaml#L126) for franka_robot_state_broadcaster to true, such that we don't return Error when we can't lock the realtime publisher and the franka_robot_state_broadcaster controller is not dropped.

BarisYazici commented:
I seem to get more `communication_constraint` errors with the Jazzy than the Humble version. Is it also the case for you? I am testing both in the `devcontainer`. Could you also test both distros and compare them? Maybe there is a ros2 control argument we are missing to pass?

Commit:
config: Change the default of lock_update_success to true

Commit:
refactor: remove duplicate code in semantic_components

Commit:
config: update .github/workflows/ci.yml

Commit:
config: Change franka_description to use the jazzy branch
